### PR TITLE
feat(SecretsManager): Add support for secrets count metric

### LINKED ISSUE #137: [secrets] Add support for secrets count metric
### Feature scope

AWS Secrets Manager

### Describe your suggested feature

Announced in https://aws.amazon.com/about-aws/whats-new/2022/05/aws-secrets-manager-publishes-usage-metrics-to-amazon-cloudwatch/

Docs: https://docs.aws.amazon.com/secretsmanager/latest/userguide/monitoring-cloudwatch.html

### PR DESCRIPTION
**Pull request by zqumei0**
Added new Secrets Manager Monitor to support Account-Level Secrets Count. This includes a new SecretsManagerMonitor, a new SecretsManagerMetricsFactory, and a new SecretsManagerAlarmFactory. There is alarm support for Min/Max Secrets count and change in secrets count.

Closes #137

---

_By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license_

### CODE REVIEW

File: lib/common/monitoring/alarms/SecretsManagerAlarmFactory.ts [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+        props.treatMissingDataOverride ?? TreatMissingData.MISSING,
+      datapointsToAlarm: props.datapointsToAlarm ?? NUMBER_OF_DATAPOINTS,
+      comparisonOperator:
+        props.comparisonOperatorOverride ??
+        ComparisonOperator.LESS_THAN_THRESHOLD,
+      ...props,
+      disambiguator,
+      threshold: props.minSecretCount,
+      alarmNameSuffix: "Secrets-Count-Min",
+      alarmDescription: "Number of secrets in the queue is too low.",
```
Comment (echeung-amzn) [inline lib/common/monitoring/alarms/SecretsManagerAlarmFactory.ts:49]:
```suggestion
      alarmDescription: "Number of secrets is too low.",
```

File: lib/common/monitoring/alarms/SecretsManagerAlarmFactory.ts [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+        props.treatMissingDataOverride ?? TreatMissingData.MISSING,
+      comparisonOperator:
+        props.comparisonOperatorOverride ??
+        ComparisonOperator.GREATER_THAN_THRESHOLD,
+      datapointsToAlarm: props.datapointsToAlarm ?? NUMBER_OF_DATAPOINTS,
+      ...props,
+      disambiguator,
+      threshold: props.maxSecretCount,
+      alarmNameSuffix: "Secrets-Count-Max",
+      alarmDescription: "Number of secrets in the queue is too high.",
```
Comment (echeung-amzn) [inline lib/common/monitoring/alarms/SecretsManagerAlarmFactory.ts:69]:
```suggestion
      alarmDescription: "Number of secrets is too high.",
```

File: lib/common/monitoring/alarms/SecretsManagerAlarmFactory.ts [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+export interface MinSecretCountThreshold extends CustomAlarmThreshold {
+  readonly minSecretCount: number;
+}
+
+export interface MaxSecretCountThreshold extends CustomAlarmThreshold {
+  readonly maxSecretCount: number;
+}
+
+export interface ChangeInSecretCountThreshold extends CustomAlarmThreshold {
+  readonly currentSecretCount: number;
```
Comment (echeung-amzn) [inline lib/common/monitoring/alarms/SecretsManagerAlarmFactory.ts:20]:
```suggestion
  readonly requiredSecretCount: number;
```

Might be a better name?

File: lib/monitoring/aws-secretsmanager/SecretsManagerMetricFactory.ts [hunk 4]
```diff
@@ -0,0 +1,37 @@
+import { Duration } from "aws-cdk-lib";
+import { MetricFactory, MetricStatistic } from "../../common";
+
+const DEFAULT_METRIC_PERIOD = Duration.hours(1);
+
+export class SecretsManagerMetricFactory {
+  static readonly Namespace = "AWS/SecretsManager";
```
Comment (echeung-amzn) [inline lib/monitoring/aws-secretsmanager/SecretsManagerMetricFactory.ts:7]: We've typically had these outside of the factory class for other similar things (i.e. at the same level as `DEFAULT_METRIC_PERIOD` here).

File: lib/monitoring/aws-secretsmanager/SecretsManagerMetricFactory.ts [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+  static readonly Type = "Resource";
+  static readonly MetricNameSecretCount = "ResourceCount";
+  protected readonly metricFactory: MetricFactory;
+
+  constructor(metricFactory: MetricFactory) {
+    this.metricFactory = metricFactory;
+  }
+
+  metricSecretCount() {
+    const dimensionsMap = {
```
Comment (echeung-amzn) [inline lib/monitoring/aws-secretsmanager/SecretsManagerMetricFactory.ts:20]: Could lift this too considering it doesn't change.

File: package-lock.json
Comment (echeung-amzn) [inline package-lock.json:1]: Revert changes to this file.

**Review by echeung-amzn — CHANGES_REQUESTED** [review]

File: lib/common/monitoring/alarms/SecretsManagerAlarmFactory.ts (hunk 3 above)
Comment (zqumei0) [reply lib/common/monitoring/alarms/SecretsManagerAlarmFactory.ts:20]: Changed the name

**Review by echeung-amzn — APPROVED** [review]

**Pull request merged**